sql/src/main/java/io/crate/executor/transport/TransportShardUpdateAction.java

```diff
@@ -0,0 +1,335 @@
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.VersionType;
+import org.elasticsearch.index.engine.DocumentMissingException;
+import org.elasticsearch.index.engine.DocumentSourceMissingException;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.mapper.internal.ParentFieldMapper;
+import org.elasticsearch.index.mapper.internal.RoutingFieldMapper;
+import org.elasticsearch.index.mapper.internal.TTLFieldMapper;
+import org.elasticsearch.index.service.IndexService;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.service.IndexShard;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.search.fetch.source.FetchSourceContext;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TransportShardUpdateAction extends TransportInstanceSingleOperationAction<ShardUpdateRequest, UpdateResponse> {
+
+    private final static Visitor VISITOR = new Visitor();
+
+    private final TransportIndexAction indexAction;
+    private final IndicesService indicesService;
+    private final Functions functions;
+
+    @Inject
+    public TransportShardUpdateAction(Settings settings,
+                                      String actionName,
+                                      ThreadPool threadPool,
+                                      ClusterService clusterService,
+                                      TransportService transportService,
+                                      ActionFilters actionFilters,
+                                      TransportIndexAction indexAction,
+                                      IndicesService indicesService,
+                                      Functions functions) {
+        super(settings, actionName, threadPool, clusterService, transportService, actionFilters);
+        this.indexAction = indexAction;
+        this.indicesService = indicesService;
+        this.functions = functions;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.INDEX;
+    }
+
+    @Override
+    protected ShardUpdateRequest newRequest() {
+        return new ShardUpdateRequest();
+    }
+
+    @Override
+    protected UpdateResponse newResponse() {
+        return new UpdateResponse();
+    }
+
+    @Override
+    protected boolean retryOnFailure(Throwable e) {
+        return TransportActions.isShardNotAvailableException(e);
+    }
+
+    @Override
+    protected boolean resolveRequest(ClusterState state, InternalRequest request, ActionListener<UpdateResponse> listener) {
+        request.request().routing((state.metaData().resolveIndexRouting(request.request().routing(), request.request().index())));
```

Comment:
could you delegate the resolveRequest stuff to the class where you copied the logic from`
 So you would just call  `updateAction.resolveRequest(...)` or something like that here? Not sure if it works if we have a different UpdateResponse class